### PR TITLE
Shrestha/backend manager fixes

### PR DESCRIPTION
**Pull request by shresthamalik**
Changes: 
-  IsSupportedBackend tries to create a backend rather than querying nGraph for registered backends. There were some error conditions arising due to static initialization of the supported_backends map and late registration of backends, which led to this change
- GetCurrentlySetBackendName accounts for the NGRAPH_TF_Backend set
- Update BackendManager Tests

### CODE REVIEW

File: src/ngraph_backend_manager.h [hunk 1]
```diff
@@ -47,25 +47,25 @@ struct Backend {
 class BackendManager {
  public:
   // Returns the backend name currently set
-  static string GetCurrentlySetBackendName() {
-    return BackendManager::ng_backend_name_;
-  };
+  // If env variable NGRAPH_TF_BACKEND is set it has precedence
```
Comment (kanvi-nervana) [inline src/ngraph_backend_manager.h]: It has precedence only over the BackendManager and not the config, correct?

File: test/python/common.py [hunk 2]
```diff
@@ -91,3 +91,29 @@ def generate_random_numbers(self,
         if datatype == "DTYPE_INT":
             return [random.randint(start, end) for i in range(vector_length)]
         return [random.uniform(start, end) for i in range(vector_length)]
+
+    # returns true if the the env variable is set
+    def is_env_variable_set(self, env_var):
+        return env_var in os.environ
+
+    # sets the env variable
+    def set_env_variable(self, env_var, env_var_val):
```
Comment (kanvi-nervana) [inline test/python/common.py]: Is this only for setting the backend env variable? If yes, then please change the name to project that.
If not, then please change the `print message`

File: test/python/test_set_backend.py [hunk 3]
```diff
@@ -19,54 +19,68 @@
...
-    ngraph_bridge.enable()
-    backend_cpu = 'CPU'
-    backend_interpreter = 'INTERPRETER'
+class TestSetBackend(NgraphTest):
+
+    def test_set_backend(self):
+        # store env variables
+        # when testing on backends like GPU the tests are run with NGRPAH_TF_BACKEND
+        # by storing and restoring the env_variables we run the tests independent of the backend set
+        # currently we only and restore only NGRPAH_TF_BACKEND
```
Comment (kanvi-nervana) [inline test/python/test_set_backend.py]: typo: ` # currently we only and restore only NGRPAH_TF_BACKEND` ---> ` # currently we only restore NGRPAH_TF_BACKEND`

File: src/ngraph_backend_manager.cc [hunk 4]
```diff
@@ -101,22 +107,52 @@ void BackendManager::UnlockBackend(const string& backend_name) {
...
+
+  // NGRAPH_TF_BACKEND is set
+  string backend_env = std::string(ng_backend_env_value);
+  if (backend_env.empty() || !BackendManager::IsSupportedBackend(backend_env)) {
+    return errors::Internal("NGRAPH_TF_BACKEND: ", backend_env,
+                            " is not supported");
+  }
+
+  *backend_name = backend_env;
+  NGRAPH_VLOG(1) << "Overriding backend using the enviornment variable "
```
Comment (sayantan-nervana) [inline src/ngraph_backend_manager.cc]:
```suggestion
  NGRAPH_VLOG(1) << "Overriding backend using the environment variable "
```

File: src/ngraph_backend_manager.h (hunk 1 above)
Comment (shresthamalik) [reply src/ngraph_backend_manager.h]: Yes.

File: test/python/common.py (hunk 2 above)
Comment (shresthamalik) [reply test/python/common.py]: This is for setting any env variable.

**Pull request merged**